src/pymmcore_widgets/_mda/_positions_table_widget.py

```diff
@@ -213,88 +267,217 @@ def _add_position_row(self) -> int:
+        add_grid.setIconSize(QSize(25, 25))
+        add_grid.setFixedHeight(25)
+        add_grid.setSizePolicy(QSizePolicy.Policy.Fixed, QSizePolicy.Policy.Fixed)
+        add_grid.setContextMenuPolicy(Qt.CustomContextMenu)
+        # for righ-click menu
+        add_grid.customContextMenuRequested.connect(self._show_apply_to_all_menu)
+        add_grid.clicked.connect(self._grid_widget)
+        remove_grid = QPushButton()
+        remove_grid.setIcon(icon(MDI6.close_thick, color="magenta"))
+        remove_grid.setIconSize(QSize(25, 25))
+        remove_grid.setFixedHeight(25)
+        remove_grid.setSizePolicy(QSizePolicy.Policy.Fixed, QSizePolicy.Policy.Fixed)
+        remove_grid.clicked.connect(self._remove_grid_plan)
+        layout.addWidget(add_grid)
+        layout.addWidget(remove_grid)
+        remove_grid.hide()
+        self._table.setCellWidget(row, col, wdg)
+
+    def _remove_grid_plan(self) -> None:
+        row = self._table.indexAt(self.sender().parent().pos()).row()
+        self._table.item(row, 0).setData(self.GRID_ROLE, None)
+        self._table.item(row, 0).setToolTip("")
+        add_grid, remove_grid = self._get_grid_buttons(row)
+        add_grid.setText("")
+        add_grid.setIcon(icon(MDI6.plus_thick, color=(0, 255, 0)))
+        add_grid.setIconSize(QSize(25, 25))
+        remove_grid.hide()
+        self._enable_button()
+        self.valueChanged.emit()
 
-    def _remove_position(self) -> None:
-        rows = {r.row() for r in self._table.selectedIndexes()}
-        grid_to_delete = []
-
-        for idx in sorted(rows, reverse=True):
-            pos_role = self._table.item(idx, 0).data(self.POS_ROLE)
-            # store grid name if is a grid position
-            if "Grid" in pos_role:
-                grid_name = pos_role.split("_")[0]
-                grid_to_delete.append(grid_name)
-            else:
-                # remove if is a single position
-                self._table.removeRow(idx)
+    def _get_grid_buttons(self, row: int) -> tuple[QPushButton, QPushButton]:
+        return (
+            self._table.cellWidget(row, 4).layout().itemAt(0).widget(),
+            self._table.cellWidget(row, 4).layout().itemAt(1).widget(),
+        )
+
+    def _grid_widget(self) -> None:
+        if not self._mmc.getXYStageDevice():
+            return
 
-        # remove grid positions
-        for gridname in grid_to_delete:
-            self._delete_grid_positions(gridname)
+        if hasattr(self, "_grid_wdg"):
+            self._grid_wdg.close()  # type: ignore
 
-        self._rename_positions()
+        self._grid_wdg = GridWidget(
+            parent=self,
+            mmcore=self._mmc,
+            current_stage_pos=(self._mmc.getXPosition(), self._mmc.getYPosition()),
+        )
+        row = self._table.indexAt(self.sender().parent().pos()).row()
+        self._grid_wdg.valueChanged.connect(lambda x: self._add_grid_plan(x, row))
```

Comment:
this is a bit of a circuitous way to use a Dialog (i.e. you're not taking advantage of what a QDialog does). 

You have a button ("add grid") that essentially "accepts" the dialog, but then emits a signal, and the _other_ widget is hooked up to that value emission and does the job of closing the dialog.

Modal QDialogs have a whole structure for doing exactly this (see discussion of `accept`  and `reject` in https://doc.qt.io/qt-6/qdialog.html#modal-dialogs ...